File: PiDay/Core/Domain/PiMatch.swift
import Foundation

// WHY: All pi-search result types live here. They form the core "output" of the
// repository layer — both the ViewModel and Views depend on these shapes.

struct PiMatchResult: Identifiable, Equatable {
    let query: String
    let format: DateFormatOption
    let found: Bool
    let storedPosition: Int?
    let excerpt: String?

    var id: String { "\(format.rawValue)-\(query)" }

    static func sortByPosition(_ left: PiMatchResult, _ right: PiMatchResult) -> Bool {
        switch (left.storedPosition, right.storedPosition) {
        case let (lhs?, rhs?): return lhs < rhs
        case (_?, nil):        return true
        case (nil, _?):        return false
        case (nil, nil):       return left.format.displayName < right.format.displayName
        }
    }
}

// BestPiMatch is always a found result with a confirmed position and excerpt.
// Keeping it separate from PiMatchResult avoids Optional gymnastics at the call site.
struct BestPiMatch: Equatable {
    let format: DateFormatOption
    let query: String
    let storedPosition: Int
    let excerpt: String

    // WHY: When multiple formats match, any zero-padded format (e.g. DDMMYYYY "03052026")
    // is preferred over D/M/YYYY without leading zeros (e.g. "352026"), even if the
    // no-leading-zeros sequence appears earlier in pi. Shorter strings match more
    // frequently in any digit stream simply because there are more candidate windows;
    // users recognise a padded date string as their birthday — a bare "352026" looks
    // like a bug (leading zeros "dropped"). Only fall back to dmyNoLeadingZeros when
    // no padded format produced a match.
    static func preferringPadded(_ lhs: BestPiMatch, _ rhs: BestPiMatch) -> Bool {
        let lhsPadded = lhs.format != .dmyNoLeadingZeros
        let rhsPadded = rhs.format != .dmyNoLeadingZeros
        if lhsPadded != rhsPadded { return lhsPadded }
        return lhs.storedPosition < rhs.storedPosition
    }
}

// Where the result came from — bundled JSON or the live API.
enum LookupSource: String, Equatable {
    case bundled
    case live

    var label: String {
        switch self {
        case .bundled:
            return "Bundled"
        case .live:
            return "Live"
        }
    }
}

struct DateLookupSummary: Equatable {
    let isoDate: String
    let matches: [PiMatchResult]
    let bestMatch: BestPiMatch?
    let source: LookupSource
    let errorMessage: String?

    var foundCount: Int {
        matches.filter(\.found).count
    }
}

File: PiDayAndroid/app/src/main/java/academy/glasscode/piday/core/domain/PiMatch.kt
package academy.glasscode.piday.core.domain

data class PiMatchResult(
    val query: String,
    val format: DateFormatOption,
    val found: Boolean,
    val storedPosition: Int?,
    val excerpt: String?
) {
    companion object {
        // WHY: Matches with a position sort before those without; ties sort by format name.
        val byPosition: Comparator<PiMatchResult> = Comparator { l, r ->
            when {
                l.storedPosition != null && r.storedPosition != null -> l.storedPosition - r.storedPosition
                l.storedPosition != null -> -1
                r.storedPosition != null ->  1
                else -> l.format.displayName.compareTo(r.format.displayName)
            }
        }
    }
}

data class BestPiMatch(
    val format: DateFormatOption,
    val query: String,
    val storedPosition: Int,
    val excerpt: String
) {
    companion object {
        // WHY preferPadded: shorter strings (D/M/YYYY) match more often by chance.
        // Users recognise "14032026" as their birthday; "1432026" looks like a bug.
        val preferringPadded: Comparator<BestPiMatch> = Comparator { l, r ->
            val lPadded = l.format != DateFormatOption.DMY_NO_LEADING_ZEROS
            val rPadded = r.format != DateFormatOption.DMY_NO_LEADING_ZEROS
            when {
                lPadded != rPadded -> if (lPadded) -1 else 1
                else               -> l.storedPosition - r.storedPosition
            }
        }
    }
}

enum class LookupSource { BUNDLED, LIVE }

data class DateLookupSummary(
    val isoDate: String,
    val matches: List<PiMatchResult>,
    val bestMatch: BestPiMatch?,
    val source: LookupSource,
    val errorMessage: String?
) {
    val foundCount: Int get() = matches.count { it.found }
}
